server/bleep/src/webserver/aaa.rs

```diff
@@ -0,0 +1,263 @@
+use std::{
+    sync::Arc,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use crate::{
+    remotes::{self, BackendCredential},
+    state::Backend,
+};
+
+use super::*;
+use anyhow::{Context, Result};
+use axum::{
+    extract::Query,
+    http::{Request, StatusCode},
+    middleware::Next,
+    response::Redirect,
+};
+use axum_extra::extract::cookie::{Cookie, CookieJar};
+use dashmap::DashSet;
+use rand::{distributions::Alphanumeric, Rng};
+use secrecy::ExposeSecret;
+use serde::Deserialize;
+use tantivy::time::Duration;
+use tracing::error;
+
+const AUTH_TOKEN_COOKIE: &str = "auth_token";
+const AUTH_CREATED_AT_COOKIE: &str = "auth_created_at";
+
+const STATE_LEN: usize = 32;
+type State = String;
+
+/// Initiate a new login using a web-based OAuth flow.
+///
+/// Redirects to Github after setting up the correct expectations.
+/// Use this endpoint as a redirect target instead of an API call!
+#[utoipa::path(get, path = "/auth/login/start",
+    responses(
+        (status = 200, description = "Execute query successfully", body = Response),
+        (status = 400, description = "Bad request", body = EndpointError),
+        (status = 500, description = "Server error", body = EndpointError),
+    ),
+)]
+pub(super) async fn login(
+    Extension(app): Extension<Application>,
+    Extension(auth_layer): Extension<Arc<AuthLayer>>,
+) -> impl IntoResponse {
+    let state = rand::thread_rng()
+        .sample_iter(Alphanumeric)
+        .take(STATE_LEN)
+        .map(|c| c as char)
+        .collect::<String>();
+
+    let client_id = app
+        .config
+        .github_client_id
+        .as_ref()
+        .expect("github client id must be provided")
+        .expose_secret();
+
+    let redirect_uri = format!(
+        "https://{}/auth/login/complete",
+        app.config
+            .instance_domain
+            .as_ref()
+            .expect("instance domain must be provided")
+    );
+
+    let github_oauth_url = &format!(
+        "https://github.com/login/oauth/authorize\
+         ?client_id={client_id}\
+         &state={state}\
+         &allow_signup=false\
+         &redirect_uri={redirect_uri}",
+    );
+
+    auth_layer.initialized_login.insert(state);
+
+    Redirect::to(github_oauth_url)
```

Comment:
I don't think redirects will work for us, we should find another way of doing it. When a client makes a request to this endpoint and gets a redirect in response it tried to make an API call to this address, but it should open it as a webpage in a new tab instead. Also it won't work in Tauri app.

Reply:
Updated to return `{ "oauth_url": "https://..." }` to make it easier.

Reply:
thanks